### PR TITLE
MM-21307 Get/Update/Delete user preferences Plugin API

### LINKED ISSUE #13940: Extend plugin API to get/update/delete user preferences
Extend the plugin API with the following:

```go
GetPreferencesForUser(userId string) ([]model.Preference, *model.AppError)
UpdatePreferencesForUser(userId string, []model.Preference) *model.AppError
DeletePreferencesForUser(userId string, []model.Preference) *model.AppError
```

----

If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-21307

### PR DESCRIPTION
**Pull request by NassimBounouas**
#### Summary

This PR resolves #13940 by implementing Get / Update / Delete user preferences from the plugin API

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-21307

### CODE REVIEW

File: app/plugin_api_test.go [hunk 1]
```diff
@@ -130,6 +130,103 @@ func TestPublicFilesPathConfiguration(t *testing.T) {
...
+		Username: "user1" + model.NewId(),
+	})
+	require.Nil(t, err)
+	defer th.App.PermanentDeleteUser(user1)
+
+	preferences, err := api.GetPreferencesForUser(user1.Id)
+	require.Nil(t, err)
+	assert.Equal(t, 1, len(preferences))
+
+	err = api.DeletePreferencesForUser(user1.Id, preferences)
```
[the remark this answers is not in the document]
Comment (NassimBounouas) [reply app/plugin_api_test.go]: Hi @mickmister,

Indeed that's a case I didn't cover. I will add it !

File: plugin/api.go [hunk 2]
```diff
@@ -152,6 +152,27 @@ type API interface {
 	// Minimum server version: 5.6
 	GetUsersInTeam(teamId string, page int, perPage int) ([]*model.User, *model.AppError)
 
+	// GetPreferencesForUser gets a user's preferences.
+	//
+	// @tag User
+	// @tag Preference
+	// Minimum server version: 5.22
+	GetPreferencesForUser(userId string) ([]model.Preference, *model.AppError)
```
[the remark this answers is not in the document]
Comment (NassimBounouas) [reply plugin/api.go]: I used the declaration provided in the PR but after a second look on other declarations, indeed the struct should be passed by reference. I'll do it while I'm adding the requested test case below.

[the remark this answers is not in the document]
Comment (NassimBounouas) [reply plugin/api.go]: @mickmister :

The preference implementation of the function `GetPreferencesForUser` in [app/preferences.go](https://github.com/mattermost/mattermost-server/blob/master/app/preference.go#L12) returns a `model.Preferences` which is in fact a type hiding an array :  [[]Preference](https://github.com/mattermost/mattermost-server/blob/5a34ec47935045ddc33686b0ac52384bc8a6baba/model/preferences.go#L11).

So, if I'm right it requires to modify the complete call chains to [the preference store](https://github.com/mattermost/mattermost-server/blob/5a34ec47935045ddc33686b0ac52384bc8a6baba/store/sqlstore/preference_store.go#L184) in order to reach something similar (for example) to the [GetUsersByUsernames](https://github.com/mattermost/mattermost-server/blob/master/plugin/api.go#L146) implementation which calls [GetProfilesByUsernames](https://github.com/mattermost/mattermost-server/blob/master/store/sqlstore/user_store.go#L682) from the user store.

I think it's why @lieut-data suggested those prototype without references on the issue #13940.

Can you please confirm my analysis ? If I'm right, do you want me to modify the complete call chain (which implies a lot of side effect) ?

[the remark this answers is not in the document]
Comment (lieut-data) [reply plugin/api.go]: @nassimbounouas -- I agree with your analysis, and suggest we avoid modifying the complete call chain. We can "fix" this in the mattermost plugin API wrapper instead.

File: app/plugin_api_test.go [hunk 3]
```diff
@@ -176,6 +176,41 @@ func TestPluginAPIDeleteUserPreferences(t *testing.T) {
...
+	}
+
+	err = api.DeletePreferencesForUser(user2.Id, []model.Preference{preference})
+	require.Nil(t, err)
+	preferences, err = api.GetPreferencesForUser(user2.Id)
+	require.Nil(t, err)
+	assert.Equal(t, 1, len(preferences))
+	expectedCategories = []string{model.PREFERENCE_CATEGORY_TUTORIAL_STEPS}
+	for _, pref := range preferences {
+		assert.Contains(t, expectedCategories, pref.Category)
```
Comment (mickmister) [inline app/plugin_api_test.go]: @NassimBounouas Do you mind removing the for loops in this test? It's adding extra code to the test with out much utility added. Same with the block above this one.

Thanks for adding this test case!

Comment (NassimBounouas) [reply app/plugin_api_test.go]: No problem @mickmister ! I'm doing it !

File: app/plugin_api.go [hunk 4]
```diff
@@ -247,6 +247,18 @@ func (api *PluginAPI) GetUsersInTeam(teamId string, page int, perPage int) ([]*m
...
+func (api *PluginAPI) GetPreferencesForUser(userId string) ([]model.Preference, *model.AppError) {
+	return api.app.GetPreferencesForUser(userId)
+}
+
+func (api *PluginAPI) UpdatePreferencesForUser(userId string, preferences []model.Preference) *model.AppError {
+	return api.app.UpdatePreferences(userId, preferences)
+}
+
+func (api *PluginAPI) DeletePreferencesForUser(userId string, preferences []model.Preference) *model.AppError {
+	return api.app.DeletePreferences(userId, preferences)
```
Comment (cpoile) [inline app/plugin_api.go]: As an aside, I don't understand this api...  /shrug

Comment (NassimBounouas) [reply app/plugin_api.go]: Maybe this one can be reviewed in a future PR while uniformizing the api (passing by reference) as discussed with @lieut-data

**Comment by murray484** [comment]
Any idea when we can expect to see this merged into master?

**Comment by lieut-data** [comment]
Hey @murray484, sorry for the delay! I know @prapti is hard at work helping with the release of v5.22, and should hopefully have some cycles soon to complete PR testing.

File: plugin/api.go [hunk 5]
```diff
@@ -152,6 +152,27 @@ type API interface {
 	// Minimum server version: 5.6
 	GetUsersInTeam(teamId string, page int, perPage int) ([]*model.User, *model.AppError)
 
+	// GetPreferencesForUser gets a user's preferences.
+	//
+	// @tag User
+	// @tag Preference
+	// Minimum server version: 5.22
```
Comment (lieut-data) [inline plugin/api.go:159]:
```suggestion
	// Minimum server version: 5.24
```

File: plugin/api.go [hunk 6]
```diff
@@ -152,6 +152,27 @@ type API interface {
...
+	GetPreferencesForUser(userId string) ([]model.Preference, *model.AppError)
+
+	// UpdatePreferencesForUser updates a user's preferences.
+	//
+	// @tag User
+	// @tag Preference
+	// Minimum server version: 5.22
```
Comment (lieut-data) [inline plugin/api.go:166]:
```suggestion
	// Minimum server version: 5.24
```

File: plugin/api.go [hunk 7]
```diff
@@ -152,6 +152,27 @@ type API interface {
...
+	UpdatePreferencesForUser(userId string, preferences []model.Preference) *model.AppError
+
+	// DeletePreferencesForUser deletes a user's preferences.
+	//
+	// @tag User
+	// @tag Preference
+	// Minimum server version: 5.22
```
Comment (lieut-data) [inline plugin/api.go:173]:
```suggestion
	// Minimum server version: 5.24
```

**Comment by teresa-novoa** [comment]
@prapti friendly reminder to add this PR to your review list, thanks!!

File: plugin/api.go [hunk 8]
```diff
@@ -152,6 +152,27 @@ type API interface {
 	// Minimum server version: 5.6
 	GetUsersInTeam(teamId string, page int, perPage int) ([]*model.User, *model.AppError)
 
+	// GetPreferencesForUser gets a user's preferences.
+	//
+	// @tag User
+	// @tag Preference
+	// Minimum server version: 5.24
```
Comment (saturninoabril) [inline plugin/api.go:159]: @lieut-data This feature has missed 5.24.  Should this change to 5.25?

Comment (lieut-data) [reply plugin/api.go:159]:
```suggestion
 	// Minimum server version: 5.26
```

File: plugin/api.go [hunk 9]
```diff
@@ -152,6 +152,27 @@ type API interface {
...
+	GetPreferencesForUser(userId string) ([]model.Preference, *model.AppError)
+
+	// UpdatePreferencesForUser updates a user's preferences.
+	//
+	// @tag User
+	// @tag Preference
+	// Minimum server version: 5.24
```
Comment (lieut-data) [inline plugin/api.go:166]:
```suggestion
 	// Minimum server version: 5.26 
```

File: plugin/api.go [hunk 10]
```diff
@@ -152,6 +152,27 @@ type API interface {
...
+	UpdatePreferencesForUser(userId string, preferences []model.Preference) *model.AppError
+
+	// DeletePreferencesForUser deletes a user's preferences.
+	//
+	// @tag User
+	// @tag Preference
+	// Minimum server version: 5.24
```
Comment (lieut-data) [inline plugin/api.go:173]:
```suggestion
 	// Minimum server version: 5.26 
```

**Comment by lieut-data** [comment]
Good idea, @saturninoabril! Filed https://mattermost.atlassian.net/browse/MM-25465 to track.

**Comment by lieut-data** [comment]
@NassimBounouas, before we can merge this, we'll need you to sign the CLA: https://mattermost.org/mattermost-contributor-agreement/.

**Comment by justinegeffen** [comment]
I'm following up the Plugin API docs here: https://developers.mattermost.com/extend/plugins/server/reference/#API and in the process of confirming the doc update process.

**Comment by justinegeffen** [comment]
These docs are auto-generated and will get updated when the version is bumped.

**Pull request merged**